### DISCUSSION TITLE
Wrong linking / references

### DISCUSSION THREAD

**Discussion by grafandreas** [eclipse-langium/langium#1749] [Q&A] [answered]
Creating a language with the small grammar below results in the following unexpected behaviour:

The references in line Stage: Are not actually navigable references in the generated VSCode editor
The name=ID in Microaction: will actually result in a navigable reference in VSCode editor, but it will not navigate to any definition.

```
grammar CorePerfDsl

entry CorePerfModel_sec :  (mad=Microaction_def | stdef=Stage_def)* ;

Stage_def : 'Stagxe' (st+=Stage | '{' st+=Stage (',' st+=Stage)* '}');

Stage : name=ID '(' microactions+=[Microaction] (',' microactions+=[Microaction])* ')';

//////////////////////////// MICROACTION ////////////////////////////

Microaction_def : 'Microaction' (ma+=Microaction | '{' ma+=Microaction (',' ma+=Microaction)* '}');

Microaction : name=ID '(' ')';

terminal KEYWORD_ALL  : '[ALL]' ;
terminal KEYWORD_REST : '[?]' ;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal INT returns number: /[0-9]+/;
terminal STRING: /"(\\.|[^"\\])*"|'(\\.|[^'\\])*'/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

**Comment by msujew** [comment]
Hey @grafandreas, I've converted this issue into a discussion, as you're not encountering a bug.

>  will actually result in a navigable reference in VSCode editor, but it will not navigate to any definition.

So it's a navigatable reference, but it doesn't navigate? That sounds like an oxymoron :D

Anyway, first off there's a major issue (and the latest versions of Langium should warn you about that) in your entry rule. Parsing multiple `Microaction_def` or `Stage_def` will override existing elements. You should use `+=` instead of `=` there if you want to use `*`.

Then there's the issue of linking: Langium doesn't magically know how all references are supposed to be linked. Instead, it features a relatively simplistic default mechanism (local scopes) that can be customized/extended to support your language. See our [scoping guide](https://langium.org/docs/recipes/scoping/).

**Comment by grafandreas** [comment]
@msujew It is a navigable reference in the grammar, but it is not reflected in the editor. 

Maybe I am to Xtext influenced, I was assuming that the default scoping picks up matching elements in a file (as local scope)?

**Reply by msujew** [reply, marked as answer]
Langium's default scoping uses local scopes as well. However, as `Microaction` instances are defined in the scope of a `Microaction_def`, it won't be available outside of that. That is the definition of a local scope. By default, elements are only available if they are a sibling, a parent, or a sibling of a parent (recursively). To adjust how the scoping is done, see the scoping guide posted above.
